File: app/src/main/java/pl/piotrskiba/teatime/TeaTimerFragment.java
package pl.piotrskiba.teatime;

import android.app.ActivityManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.TextView;

import com.mikhaellopez.circularprogressbar.CircularProgressBar;

import butterknife.BindView;
import butterknife.ButterKnife;
import pl.piotrskiba.teatime.services.CountDownTimerService;

public class TeaTimerFragment extends Fragment implements SeekBar.OnSeekBarChangeListener, View.OnClickListener {

    @BindView(R.id.sb_timer)
    SeekBar mTimerSeekBar;

    @BindView(R.id.tv_seekbar_title)
    TextView mTimerSeekBarTitle;

    @BindView(R.id.pb_timer)
    CircularProgressBar mTimerProgressBar;

    @BindView(R.id.tv_timer)
    TextView mTimerTextView;

    @BindView(R.id.btn_start_timer)
    Button mTimerStartButton;

    @BindView(R.id.btn_stop_timer)
    Button mTimerStopButton;

    @BindView(R.id.btn_disable_alarm)
    Button mDisableAlarmButton;


    private int mTeaIndex;
    private int mTotalBrewingTime;

    private TimerUpdateReceiver mTimerUpdateReceiver;

    Intent timerService;

    public boolean showAlarmLayout = false;

    public TeaTimerFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_tea_timer, container, false);

        ButterKnife.bind(this, rootView);

        populateFragment();

        return rootView;
    }

    public void setTeaIndex(int index){
        mTeaIndex = index;
    }

    private void populateFragment() {
        mTimerSeekBar.setOnSeekBarChangeListener(this);
        if(showAlarmLayout) {
            showAlarmLayout();
        }
        else if(isServiceRunning(CountDownTimerService.class)) {
            String tea_id = getResources().getStringArray(R.array.tea_ids)[mTeaIndex];
            SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getContext());
            int timeleft = sharedPreferences.getInt(getString(R.string.pref_timeleft_key, tea_id), 0);
            if(timeleft > 0){
                showBrewingLayout();

                updateTimerText(timeleft);

                int total = sharedPreferences.getInt(getString(R.string.pref_total_time_key, tea_id), 1);
                mTotalBrewingTime = total;

                int progress = (int)((float)timeleft/mTotalBrewingTime*1000);
                mTimerProgressBar.setProgress(progress);
            }
        }
        else {
            setDefaultSeekBarProgress();
        }

        mTimerStartButton.setOnClickListener(this);
        mTimerStopButton.setOnClickListener(this);
        mDisableAlarmButton.setOnClickListener(this);
    }

    private void setDefaultSeekBarProgress() {
        int max = getResources().getIntArray(R.array.tea_max_brewing_time)[mTeaIndex];

        if (max != 0) {
            int min = getResources().getIntArray(R.array.tea_min_brewing_time)[mTeaIndex];
            float defVal = getResources().getIntArray(R.array.tea_default_brewing_time)[mTeaIndex];
            int sbval = (int)((defVal-min) / (max-min) * 100);
            mTimerSeekBar.setProgress(sbval);

            onProgressChanged(mTimerSeekBar, mTimerSeekBar.getProgress(), false);
        }
    }

    private int getSeekBarValue(int progress){
        int min = getResources().getIntArray(R.array.tea_min_brewing_time)[mTeaIndex];
        int max = getResources().getIntArray(R.array.tea_max_brewing_time)[mTeaIndex];

        progress = (int)((float)progress/100 * (max-min));
        progress /= 5;
        progress *= 5;
        progress += min;

        return progress;
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        int seconds = getSeekBarValue(progress);
        updateTimerText(seconds);
    }

    private void updateTimerText(int seconds){
        int minutes = 0;

        while(seconds >= 60){
            seconds -= 60;
            minutes += 1;
        }

        mTimerTextView.setText(getString(R.string.timer_format, minutes, seconds));
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onClick(View v) {
        if(v.getId() == R.id.btn_start_timer) {
            startBrewing();
        }
        else if(v.getId() == R.id.btn_stop_timer) {
            cancelBrewing();
        }
        else if(v.getId() == R.id.btn_disable_alarm){
            throw new UnsupportedOperationException("not implemented yet");
        }
    }

    private void startBrewing(){
        mTotalBrewingTime = getSeekBarValue(mTimerSeekBar.getProgress());

        timerService = new Intent(getContext(), CountDownTimerService.class);
        timerService.putExtra(Constants.EXTRA_INDEX, mTeaIndex);
        timerService.putExtra(Constants.EXTRA_SECONDS, mTotalBrewingTime);
        getContext().getApplicationContext().startService(timerService);

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getContext().getApplicationContext());
        SharedPreferences.Editor editor = sharedPreferences.edit();
        String tea_id = getResources().getStringArray(R.array.tea_ids)[mTeaIndex];
        editor.putInt(getString(R.string.pref_total_time_key, tea_id), mTotalBrewingTime);
        editor.commit();

        showBrewingLayout();
    }

    private void cancelBrewing(){
        getContext().getApplicationContext().stopService(new Intent(getContext(), CountDownTimerService.class));
        showDefaultLayout();

        int seconds = getSeekBarValue(mTimerSeekBar.getProgress());
        updateTimerText(seconds);
    }

    public void showBrewingLayout(){
        mTimerStartButton.setVisibility(View.GONE);
        mTimerStopButton.setVisibility(View.VISIBLE);
        mTimerSeekBar.setVisibility(View.INVISIBLE);
        mTimerSeekBarTitle.setVisibility(View.INVISIBLE);
    }

    private void showDefaultLayout(){
        mTimerStartButton.setVisibility(View.VISIBLE);
        mTimerStopButton.setVisibility(View.GONE);
        mTimerSeekBar.setVisibility(View.VISIBLE);
        mTimerSeekBarTitle.setVisibility(View.VISIBLE);
        mTimerProgressBar.setProgressWithAnimation(1000);
    }

    private void showAlarmLayout(){
        updateTimerText(0);
        mTimerProgressBar.setProgress(0);

        mTimerStartButton.setVisibility(View.GONE);
        mTimerStopButton.setVisibility(View.GONE);
        mDisableAlarmButton.setVisibility(View.VISIBLE);
        mTimerSeekBar.setVisibility(View.INVISIBLE);
        mTimerSeekBarTitle.setVisibility(View.INVISIBLE);
    }

    private boolean isServiceRunning(Class serviceClass) {
        ActivityManager manager = (ActivityManager) getContext().getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void onResume() {
        super.onResume();

        if (mTimerUpdateReceiver == null)
            mTimerUpdateReceiver = new TimerUpdateReceiver();
        IntentFilter intentFilter = new IntentFilter(Constants.TIMER_UPDATE_ACTION);
        getContext().registerReceiver(mTimerUpdateReceiver, intentFilter);
    }

    @Override
    public void onPause() {
        super.onPause();

        if(mTimerUpdateReceiver != null)
            getContext().unregisterReceiver(mTimerUpdateReceiver);
    }

    private class TimerUpdateReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals(Constants.TIMER_UPDATE_ACTION)) {
                int teaIndex = intent.getIntExtra(Constants.EXTRA_INDEX, -1);
                if(teaIndex == mTeaIndex) {
                    int seconds = intent.getIntExtra(Constants.EXTRA_SECONDS, -1);
                    updateTimerText(seconds);

                    int progress = (int)((float)seconds/mTotalBrewingTime*1000);
                    mTimerProgressBar.setProgress(progress);
                }
            }
        }
    }
}

File: app/src/main/java/pl/piotrskiba/teatime/TeaTimerFragment.kt
package pl.piotrskiba.teatime

import android.app.ActivityManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import butterknife.BindView
import butterknife.ButterKnife
import com.mikhaellopez.circularprogressbar.CircularProgressBar
import pl.piotrskiba.teatime.services.CountDownTimerService

class TeaTimerFragment : Fragment(), OnSeekBarChangeListener, View.OnClickListener {

    @BindView(R.id.sb_timer)
    lateinit var mTimerSeekBar: SeekBar

    @BindView(R.id.tv_seekbar_title)
    lateinit var mTimerSeekBarTitle: TextView

    @BindView(R.id.pb_timer)
    lateinit var mTimerProgressBar: CircularProgressBar

    @BindView(R.id.tv_timer)
    lateinit var mTimerTextView: TextView

    @BindView(R.id.btn_start_timer)
    lateinit var mTimerStartButton: Button

    @BindView(R.id.btn_stop_timer)
    lateinit var mTimerStopButton: Button

    @BindView(R.id.btn_disable_alarm)
    lateinit var mDisableAlarmButton: Button

    private var mTeaIndex = 0
    private var mTotalBrewingTime: Long = 0
    private var mTimerUpdateReceiver: TimerUpdateReceiver? = null

    var showAlarmLayout = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_tea_timer, container, false)

        ButterKnife.bind(this, rootView)

        if (savedInstanceState != null && savedInstanceState.containsKey(Constants.EXTRA_INDEX)) {
            mTeaIndex = savedInstanceState.getInt(Constants.EXTRA_INDEX)
        }

        return rootView
    }

    fun setTeaIndex(index: Int) {
        mTeaIndex = index
    }

    private fun populateFragment() {
        mTimerSeekBar.setOnSeekBarChangeListener(this)

        if (showAlarmLayout) {
            showAlarmLayout()
        } else if (isServiceRunning) {
            val teaId = resources.getStringArray(R.array.tea_ids)[mTeaIndex]
            val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

            val timeleft = sharedPreferences.getLong(getString(R.string.pref_timeleft_key, teaId), 0)
            if (timeleft > 0) {
                setDefaultSeekBarProgress()
                showBrewingLayout()
                updateTimerText(timeleft)

                mTotalBrewingTime = sharedPreferences.getLong(getString(R.string.pref_total_time_key, teaId), 1)
                val progress = (timeleft.toFloat() / mTotalBrewingTime * 1000).toInt()
                mTimerProgressBar.progress = progress.toFloat()
            } else {
                setDefaultSeekBarProgress()
            }
        } else {
            setDefaultSeekBarProgress()
        }

        mTimerStartButton.setOnClickListener(this)
        mTimerStopButton.setOnClickListener(this)
        mDisableAlarmButton.setOnClickListener(this)
    }

    private fun setDefaultSeekBarProgress() {
        val max = resources.getIntArray(R.array.tea_max_brewing_time)[mTeaIndex]
        if (max != 0) {
            val min = resources.getIntArray(R.array.tea_min_brewing_time)[mTeaIndex]
            val defVal = resources.getIntArray(R.array.tea_default_brewing_time)[mTeaIndex].toFloat()
            val sbval = ((defVal - min) / (max - min) * 100).toInt()

            mTimerSeekBar.progress = sbval
            onProgressChanged(mTimerSeekBar, mTimerSeekBar!!.progress, false)
        }
    }

    private fun getSeekBarValue(progress: Int): Int {
        val min = resources.getIntArray(R.array.tea_min_brewing_time)[mTeaIndex]
        val max = resources.getIntArray(R.array.tea_max_brewing_time)[mTeaIndex]

        var seekBarProgress = (progress.toFloat() / 100 * (max - min)).toInt()
        seekBarProgress /= 5
        seekBarProgress *= 5
        seekBarProgress += min
        return seekBarProgress * 1000
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        val milliseconds = getSeekBarValue(progress)
        updateTimerText(milliseconds.toLong())
    }

    private fun updateTimerText(milliseconds: Long) {
        var seconds = milliseconds.toInt() / 1000
        val minutes = seconds / 60
        seconds -= minutes * 60

        mTimerTextView.text = getString(R.string.timer_format, minutes, seconds)
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {}

    override fun onStopTrackingTouch(seekBar: SeekBar) {}

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_start_timer -> {
                startBrewing()
            }
            R.id.btn_stop_timer -> {
                cancelBrewing()
            }
            R.id.btn_disable_alarm -> {
                (context as TeaDetailsActivity?)!!.stopAlarm()
                setDefaultSeekBarProgress()
                showDefaultLayout()
            }
        }
    }

    private fun startBrewing() {
        mTotalBrewingTime = getSeekBarValue(mTimerSeekBar.progress).toLong()

        val timerService = Intent(context, CountDownTimerService::class.java)
        timerService.putExtra(Constants.EXTRA_INDEX, mTeaIndex)
        timerService.putExtra(Constants.EXTRA_MILLISECONDS, mTotalBrewingTime)
        context!!.applicationContext.startService(timerService)

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context!!.applicationContext)
        val editor = sharedPreferences.edit()
        val teaId = resources.getStringArray(R.array.tea_ids)[mTeaIndex]
        editor.putLong(getString(R.string.pref_total_time_key, teaId), mTotalBrewingTime)
        editor.apply()

        showBrewingLayout()
    }

    private fun cancelBrewing() {
        context!!.applicationContext.stopService(Intent(context, CountDownTimerService::class.java))

        showDefaultLayout()

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context!!.applicationContext)
        val editor = sharedPreferences.edit()
        val teaId = resources.getStringArray(R.array.tea_ids)[mTeaIndex]
        editor.putLong(getString(R.string.pref_timeleft_key, teaId), 0)
        editor.apply()

        val seconds = getSeekBarValue(mTimerSeekBar.progress)
        updateTimerText(seconds.toLong())
    }

    private fun showBrewingLayout() {
        mTimerStartButton.visibility = View.GONE
        mTimerStopButton.visibility = View.VISIBLE
        mDisableAlarmButton.visibility = View.GONE
        mTimerSeekBar.visibility = View.INVISIBLE
        mTimerSeekBarTitle.visibility = View.INVISIBLE
    }

    private fun showDefaultLayout() {
        mTimerStartButton.visibility = View.VISIBLE
        mTimerStopButton.visibility = View.GONE
        mDisableAlarmButton.visibility = View.GONE
        mTimerSeekBar.visibility = View.VISIBLE
        mTimerSeekBarTitle.visibility = View.VISIBLE
        mTimerProgressBar.setProgressWithAnimation(1000f)
    }

    private fun showAlarmLayout() {
        updateTimerText(0)
        mTimerProgressBar.progress = 0f
        mTimerStartButton.visibility = View.GONE
        mTimerStopButton.visibility = View.GONE
        mDisableAlarmButton.visibility = View.VISIBLE
        mTimerSeekBar.visibility = View.INVISIBLE
        mTimerSeekBarTitle.visibility = View.INVISIBLE
    }

    private val isServiceRunning: Boolean
        get() {
            val manager = context!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            for (service in manager.getRunningServices(Int.MAX_VALUE)) {
                if (CountDownTimerService::class.java.name == service.service.className) {
                    return true
                }
            }
            return false
        }

    override fun onResume() {
        super.onResume()

        if (mTimerUpdateReceiver == null)
            mTimerUpdateReceiver = TimerUpdateReceiver()

        val intentFilter = IntentFilter(Constants.TIMER_UPDATE_ACTION)
        context!!.registerReceiver(mTimerUpdateReceiver, intentFilter)

        inForeground = true

        populateFragment()
    }

    override fun onPause() {
        super.onPause()

        if (mTimerUpdateReceiver != null)
            context!!.unregisterReceiver(mTimerUpdateReceiver)

        inForeground = false
    }

    private inner class TimerUpdateReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == Constants.TIMER_UPDATE_ACTION) {
                val teaIndex = intent.getIntExtra(Constants.EXTRA_INDEX, -1)
                if (teaIndex == mTeaIndex) {
                    val milliseconds = intent.getLongExtra(Constants.EXTRA_MILLISECONDS, 0)

                    if (milliseconds == 0L) {
                        showAlarmLayout()
                        (getContext() as TeaDetailsActivity?)!!.startAlarm()
                    } else {
                        updateTimerText(milliseconds)
                        val progress = (milliseconds.toFloat() / mTotalBrewingTime * 1000).toInt()
                        mTimerProgressBar.progress = progress.toFloat()
                    }
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putInt(Constants.EXTRA_INDEX, mTeaIndex)
    }

    companion object {
        @JvmField
        var inForeground = true
    }
}